File: leetcode-new/39.cpp
class Solution
{
public:
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        vector<int> cand;
        vector<vector<int>> res;
        dfs(0, 0, cand, res, candidates, target);
        return res;
    }

    void dfs(int sum, int i, vector<int> &cand, vector<vector<int>> &res, vector<int> &candidates, int target)
    {
        if (sum > target)
            return;
        if (sum == target)
        {
            res.push_back(cand);
            return;
        }
        if (i >= candidates.size())
            return;
        cand.push_back(candidates[i]);
        dfs(sum + candidates[i], i, cand, res, candidates, target);
        cand.pop_back();
        dfs(sum, i + 1, cand, res, candidates, target);
    }
};

File: leetcode-venture/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.res = []
        def backtrack(cand, sum):
            if sum == target:
                self.res.append(cand)
                return
            if sum > target:
                return
            for i in candidates:
                if cand[-1] <= i:
                    backtrack(cand+[i], sum+i)
        
        for i in candidates:
            backtrack([i], i)
        return self.res
